sqlglot/optimizer/simplify.py

```diff
@@ -369,6 +370,51 @@ def absorb_and_eliminate(expression, root=True):
     return expression
 
 
+def propagate_constants(expression, root=True):
+    """
+    Propagate constants for conjunctions in DNF:
+
+    SELECT * FROM t WHERE a = b AND b = 5 becomes
+    SELECT * FROM t WHERE a = 5 AND b = 5
+
+    Reference: https://www.sqlite.org/optoverview.html
+    """
+    from sqlglot.optimizer.normalize import normalized
+
+    if (
+        isinstance(expression, exp.And)
+        and (root or not expression.same_parent)
+        and normalized(expression, dnf=True)
+    ):
+        constant_mapping: t.Dict[exp.Column, t.Tuple[int, exp.Literal]] = {}
+        for eq in expression.find_all(exp.EQ):
+            l, r = eq.left, eq.right
+
+            if isinstance(l, exp.Column) and isinstance(r, exp.Literal):
+                pass
+            elif isinstance(r, exp.Column) and isinstance(l, exp.Literal):
+                l, r = r, l
+            else:
+                continue
+
+            constant_mapping[l] = (id(l), r)
+
+        if constant_mapping:
+            for column in expression.find_all(exp.Column):
```

Comment:
Be careful with `expression.find_all` in predicates - this will traverse into subqueries. I think you want [find_all_in_scope](https://github.com/tobymao/sqlglot/blob/main/sqlglot/optimizer/scope.py#L777)

Reply:
Good catch, I'll make the change

Reply:
Fixed - added a test case for this too.

Reply:
I found another subtle bug with `find_all` vs `find_all_in_scope`, see https://github.com/tobymao/sqlglot/pull/2386/files#diff-18e3fc3b5cad74391bd8f9f25228b86f02a446bbd9dbecbb3d9083e8c565c2fdR87. The corresponding test case is https://github.com/tobymao/sqlglot/pull/2386/files#diff-93a74220acfd05f3226eaa0df3befb5b86cd4c8bfed680556dbfd29ce23e75cfR904-R905